rasterio/merge.py

```diff
@@ -392,3 +397,317 @@ def nullcontext(obj):
+                        / output_transform.e
+                    ),
+                    xSize=str(src.width * src.transform.a / output_transform.a),
+                    ySize=str(src.height * src.transform.e / output_transform.e),
+                )
+
+                if src.nodata is not None:
+                    ET.SubElement(complexsource, "NODATA").text = str(src.nodata)
+
+                if src.options is not None:
+                    openoptions = ET.SubElement(complexsource, "OpenOptions")
+                    for ookey, oovalue in src.options.items():
+                        ET.SubElement(openoptions, "OOI", key=str(ookey)).text = str(
+                            oovalue
+                        )
+
+    if all(MaskFlags.per_dataset in flags for flags in mask_flag_enums):
+        maskband = ET.SubElement(vrtdataset, "MaskBand")
+        vrtrasterband = ET.SubElement(maskband, "VRTRasterBand", dataType="Byte")
+
+        for idx, dataset in enumerate(datasets):
+            with dataset_opener(dataset) as src:
+                for bidx, ci, block_shape, dtype in zip(
+                    src.indexes, src.colorinterp, src.block_shapes, src.dtypes
+                ):
+                    simplesource = ET.SubElement(vrtrasterband, "SimpleSource")
+                    ET.SubElement(
+                        simplesource, "SourceFilename", relativeToVRT="0", shared="0"
+                    ).text = _parse_path(src.name).as_vsi()
+                    ET.SubElement(simplesource, "SourceBand").text = "mask,1"
+                    ET.SubElement(
+                        simplesource,
+                        "SourceProperties",
+                        RasterXSize=str(output_width),
+                        RasterYSize=str(output_height),
+                        dataType="Byte",
+                        BlockYSize=str(block_shape[0]),
+                        BlockXSize=str(block_shape[1]),
+                    )
+                    ET.SubElement(
+                        simplesource,
+                        "SrcRect",
+                        xOff="0",
+                        yOff="0",
+                        xSize=str(src.width),
+                        ySize=str(src.height),
+                    )
+                    ET.SubElement(
+                        simplesource,
+                        "DstRect",
+                        xOff=str(
+                            (src.transform.xoff - output_transform.xoff)
+                            / output_transform.a
+                        ),
+                        yOff=str(
+                            (src.transform.yoff - output_transform.yoff)
+                            / output_transform.e
+                        ),
+                        xSize=str(src.width * src.transform.a / output_transform.a),
+                        ySize=str(src.height * src.transform.e / output_transform.e),
+                    )
+
+    return ET.tostring(vrtdataset).decode("ascii")
```

Comment:
probably utf-8 if some source filenames are not ascii ?
```suggestion
    return ET.tostring(vrtdataset).decode("utf-8")
```

Reply:
@rouault it's kind of an accident that I'm decoding... I might try to eliminate that entirely. XML is defined as sequence of bytes, so Python bytes (the default from `ET.tostring()`) makes a lot of sense. Frederik Lundh (RIP), the author, was adamant about this.

I've been convert to a Python `str` because `rasterio.open` expected a `str`, not `bytes`. But I should just change the function to accept an XML byte string.